Add Airzone diagnostics

Noltari opened the pull request:
## Proposed change
Add Airzone diagnostics.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Noltari committed:
airzone: add diagnostics support

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tests: airzone: add missing typing

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
Maybe it would be better to test some of the most important coordinator data values instead of adding a full fixture which will need updating as the aioairzone library evolves.
Is that allowed?

Kane610 commented:
> Maybe it would be better to test some of the most important coordinator data values instead of adding a full fixture which will need updating as the aioairzone library evolves.
> Is that allowed?

I don't think there is a requirement to have test coverage of diagnostics so should be fine, it all depends on how much help you want with the tests when doing changes to the library.

Noltari committed:
tests: airzone: remove diagnostics fixture

Only check some key values instead of the full diagnostics, since it would
increase the changes whenever the aioairzone library evolved.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
> > Maybe it would be better to test some of the most important coordinator data values instead of adding a full fixture which will need updating as the aioairzone library evolves.
> > Is that allowed?
> 
> I don't think there is a requirement to have test coverage of diagnostics so should be fine, it all depends on how much help you want with the tests when doing changes to the library.

Thanks for the info @Kane610,
I've decided to remove the full fixture and only check for some of the values.
This will reduce the number of changes in the future when the aioairzone library evolves.

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-diagnostics

Noltari committed:
airzone: diagnostics: use correct constant

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-diagnostics

bdraco commented on homeassistant/components/airzone/diagnostics.py:
```diff
@@ -0,0 +1,31 @@
+    """Return diagnostics for a config entry."""
+    coordinator: AirzoneUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    diagnostics_data = {
+        "info": async_redact_data(config_entry.data, TO_REDACT),
+        "data": async_redact_data(coordinator.data, TO_REDACT),
+    }
+
+    return diagnostics_data
```
```suggestion
    return {
        "info": async_redact_data(config_entry.data, TO_REDACT),
        "data": async_redact_data(coordinator.data, TO_REDACT),
    }
```

Noltari committed:
airzone: diagnostics: add changes suggested by @bdraco

Co-authored-by: J. Nick Koston <nick@koston.org>

Noltari commented:
Thanks @bdraco! 🎉
